deliver/lib/deliver/upload_screenshots.rb

```diff
@@ -9,23 +31,52 @@ def upload(options, screenshots)
       v = app.edit_version
       UI.user_error!("Could not find a version to edit for app '#{app.name}'") unless v
 
-      UI.message("Starting with the upload of screenshots...")
+      @checksums_local = ScreenshotMD5s.new
+      @checksums_remote = ScreenshotMD5s.new
+
+      screenshots.each do |screenshot|
+        md5 = Spaceship::Utilities.md5digest(screenshot.path)
+        @checksums_local.add_md5(screenshot.language, screenshot.device_type, md5)
+      end
+
+      v.screenshots.each do |lang, screenshots_for_lang|
+        screenshots_for_lang.each do |current|
+          original_file_name = current.original_file_name
+          matched = original_file_name.match(/ftl_([0-9a-f]{32})_(.*)/)
 
-      # First, clear all previously uploaded screenshots, but only where we have new ones
-      # screenshots.each do |screenshot|
-      #   to_remove = v.screenshots[screenshot.language].find_all do |current|
-      #     current.device_type == screenshot.device_type
-      #   end
-      #   to_remove.each { |t| t.reset! }
-      # end
+          if matched
+            md5 = matched[1]
+            original_file_name = matched[2]
+          end
+          # store remote checksum. We will need it later to determine if we have to upload screenshot
+          @checksums_remote.add_md5(current.language, current.device_type, md5)
+          # Remove from ITC non existing locally screenshots
+          # Or screenshots that have wrong order (compare indexes)
+          # Or screenshots that we haven't uploaded through spaceship
+          unless md5 && @checksums_local.matches_md5?(current.language, current.device_type, md5, current.sort_order)
+            UI.message("Deleting screenshot #{original_file_name} for language #{current.language}")
```

Comment:
we missed that one

I would add here `##{current.sort_order}` (to display the index)